### PR TITLE
Constrained token inference interface

### PR DESCRIPTION
**Pull request by VictorOdede**
PR for #106
- Added generic `TokenConstraints` class with crude token filter
- Implemented regex to dfa compiler
- Implemented biased vocab mask for logits processor
- Added transformer class for constrained generation
- Implemented regex constraint class

WIP:
- Implementing cfg parser
- filtering valid tokens using dfa
- Adding tests

### CODE REVIEW

**Comment by cartazio** [comment]
@abhigya-sodani  is leading a new thing in response to that recent paper that makes a massive improvement in the algorithmics in this space, hit him up to see how to help out

**Comment by VictorOdede** [comment]
> @abhigya-sodani is leading a new thing in response to that recent paper that makes a massive improvement in the algorithmics in this space, hit him up to see how to help out

Great!! Would love to help with the implementation. Will definitely reach out to @abhigya-sodani
